src/main/java/net/minecraftforge/server/permission/nodes/PermissionNode.java

```diff
@@ -0,0 +1,162 @@
+import net.minecraftforge.server.permission.handler.IPermissionHandler;
+
+/**
+ * Represents the basic unit at the heart of the permission system.
+ *
+ * <p>A permission indicates the ability for an actor to perform an action,
+ * in its most general sense. In the permission system, all permissions are encoded as instances of this class,
+ * optionally integrated by a {@link PermissionDynamicContext}.</p>
+ *
+ * <p>A node is uniquely identified by its `nodeName`,
+ * which is a dot-separated string providing meaning to the node itself.
+ * The suggested structure of the name is `modid.name`, where `modid` is the ID of the owner of the node.
+ * This API <strong>does not require</strong> any implicit hierarchy,
+ * so `modid.name` is not considered a parent of `modid.name.sub`. Such decisions are left to the {@link IPermissionHandler}.</p>
+ *
+ * <p>Each node also has an associated {@link PermissionType}, detailing its exact type,
+ * along with a {@link PermissionResolver} that is used to obtain the default value of the permission.
+ * More information can be found on their documentation.</p>
+ *
+ * <p>A node can also be bound to additional {@link PermissionDynamicContextKey}s,
+ * which are used when querying the availability of the permission.
+ * There are no restrictions on their amount. It is also not mandatory to provide a value for all dynamics in a permission query.
+ * See the above link for more information.</p>
+ *
+ * <p>Each node should be registered via the {@link PermissionGatherEvent.Nodes} and stored statically in a field.
+ * That instance should then be reused every-time a permission check needs to be performed via {@link net.minecraftforge.server.permission.PermissionAPI#getPermissions(ServerPlayer, PermissionNode, PermissionDynamicContext[])}.</p>
+ */
+public final class PermissionNode<T>
+{
+   private final String nodeName;
+   private final PermissionType<T> type;
+   private final PermissionResolver<T> resolver;
+   private final PermissionDynamicContextKey<?>[] dynamics;
+
+   private String readableName;
+   private String description;
+   private LazyOptional<Pair<String, String>> translationKeys;
+
+   /**
+    * Calls {@link PermissionNode#PermissionNode(String, PermissionType, PermissionResolver, PermissionDynamicContextKey[])}
+    * with "namespace.path" as the first parameter
+    */
+   public PermissionNode(ResourceLocation nodeName, PermissionType<T> type, PermissionResolver<T> resolver, PermissionDynamicContextKey... dynamics)
+   {
+      this(nodeName.getNamespace() + "." + nodeName.getPath(), type, resolver, dynamics);
+   }
+
+   /**
+    * Calls {@link PermissionNode#PermissionNode(String, PermissionType, PermissionResolver, PermissionDynamicContextKey[])}
+    * with "modname.nodename" as the first parameter
+    */
+   public PermissionNode(String modName, String nodeName, PermissionType<T> type, PermissionResolver<T> resolver, PermissionDynamicContextKey... dynamics)
```

Comment:
Could we change `modName` to `modId` to be more consistent with everywhere else?